### PR TITLE
Memoize unary, binary and func expressions.

### PR DESCRIPTION
**Pull request by petermattis**
Fixed a bug where "SELECT col+1 FROM table" would return an error if col
was NULL instead of evaluating to NULL.

Reorganized the scanNode initialization code so that it is mostly done
in scanNode instead of planner.Select.

### CODE REVIEW

File: sql/scan.go [hunk 1]
```diff
@@ -102,24 +112,68 @@ func (n *scanNode) Err() error {
...
-// scan. Initialization consists of replacing QualifiedName nodes with
-// ParenExpr nodes for which the wrapped expression can be changed for each
-// row.
-func (n *scanNode) initExprs() bool {
-	n.qvals = make(qvalMap)
-	for i := range n.render {
-		n.render[i], n.err = n.extractQVals(n.render[i])
-		if n.err != nil {
-			return false
+	case 1:
```
Comment (tamird) [inline sql/scan.go]: this looks like a lot of code movement that is unrelated to the "real" change in this PR. can it at least get its own commit?

File: sql/parser/eval.go [hunk 2]
```diff
@@ -627,12 +629,25 @@ func evalBinaryExpr(expr *BinaryExpr) (Datum, error) {
...
-	if f != nil {
-		return f(left, right)
+	ltype := reflect.TypeOf(left)
+	rtype := reflect.TypeOf(right)
+	if expr.fn == nil {
+		expr.fn = binOps[binArgs{expr.Operator, ltype, rtype}]
+		expr.ltype = ltype
+		expr.rtype = rtype
+	}
+	if expr.fn == nil {
```
Comment (tamird) [inline sql/parser/eval.go]: this is set immediately above, how can it even be nil here?

Comment (tamird) [inline sql/parser/eval.go]: oh it's a hash lookup. disregard.

File: sql/parser/eval.go [hunk 3]
```diff
@@ -667,23 +686,41 @@ func evalFuncExpr(expr *FuncExpr) (Datum, error) {
 		types = append(types, reflect.TypeOf(arg))
 	}
 
-	var impl func(DTuple) (Datum, error)
-	for _, candidate := range candidates {
-		if candidate.match(types) {
-			impl = candidate.fn
-			break
+	needMatch := true
```
Comment (tamird) [inline sql/parser/eval.go]: i think this would work just the same without needMatch

Comment (petermattis) [inline sql/parser/eval.go]: Yeah, it was just a small optimization to avoid performing the match a second time when not necessary. Removed for now.

File: sql/scan.go (hunk 1 above)
Comment (petermattis) [inline sql/scan.go]: I broke out the code movement into a separate commit. PTAL.

**Comment by tamird** [comment]
LGTM
